spatialmath/capsule.go

```diff
@@ -0,0 +1,328 @@
+	if other, ok := g.(*capsule); ok {
+		return capsuleInCapsule(c, other), nil
+	}
+	if other, ok := g.(*box); ok {
+		return capsuleInBox(c, other), nil
+	}
+	if other, ok := g.(*sphere); ok {
+		return capsuleInSphere(c, other), nil
+	}
+	if _, ok := g.(*point); ok {
+		return false, nil
+	}
+	return true, newCollisionTypeUnsupportedError(c, g)
+}
+
+// ToPointCloud converts a capsule geometry into []r3.Vector. This method takes one argument which determines
+// how many points should like on the capsule's surface. If the argument is set to 0. we automatically
+// substitute the value with defaultTotalSpherePoints.
+func (c *capsule) ToPoints(resolution float64) []r3.Vector {
+	// check for user defined spacing
+	area := 4. * math.Pi * c.radius * c.radius
+	if resolution == 0. {
+		resolution = defaultPointDensity
+	}
+	iter := area * resolution
+	if iter < defaultMinSpherePoints {
+		iter = defaultMinSpherePoints
+	}
+
+	// First points are placed on the hemisphere endcaps
+	// code taken from: https://stackoverflow.com/questions/9600801/evenly-distributing-n-points-on-a-sphere
+	// we want the number of points on the sphere's surface to grow in proportion with the sphere's radius
+	phi := math.Pi * (3.0 - math.Sqrt(5.0)) // golden angle in radians
+	var vecList []r3.Vector
+	segLen := c.length - (2 * c.radius)
+	for i := 0.; i < iter; i++ {
+		y := 1 - (i/(iter-1))*2      // y goes from 1 to -1
+		radius := math.Sqrt(1 - y*y) // radius at y
+		theta := phi * i             // golden angle increment
+		x := (math.Cos(theta) * radius) * c.radius
+		z := (math.Sin(theta) * radius) * c.radius
+		// distal hemisphere
+		if z > 0 {
+			z += segLen
+		}
+		vec := r3.Vector{x, y * c.radius, z}
+		vecList = append(vecList, vec)
+	}
+
+	// Now distribute points along the cylindrical shaft
+	totalShaftPts := (c.radius * c.length) * resolution
+	ptsPerRing := totalShaftPts / (c.length * resolution)
+	ringCnt := math.Floor(totalShaftPts / ptsPerRing)
+	zInc := c.length / (ringCnt + 1)
+	for ring := 1.; ring <= ringCnt; ring++ {
+		for ringPt := 0.; ringPt < ptsPerRing; ringPt++ {
+			theta := 2. * math.Pi * (ringPt / ptsPerRing)
+			vecList = append(vecList, r3.Vector{math.Cos(theta) * c.radius, math.Sin(theta) * c.radius, zInc * ring})
+		}
+	}
+
+	return transformPointsToPose(vecList, c.Pose())
+}
+
+func capsuleVsPointDistance(c *capsule, other r3.Vector) float64 {
+	return DistToLineSegment(c.segA, c.segB, other) - c.radius
+}
+
+func capsuleVsSphereDistance(c *capsule, other *sphere) float64 {
+	return DistToLineSegment(c.segA, c.segB, other.pose.Point()) - (c.radius + other.radius)
+}
+
+func capsuleVsCapsuleDistance(c, other *capsule) float64 {
+	return SegmentDistanceToSegment(c.segA, c.segB, other.segA, other.segB) - (c.radius + other.radius)
+}
+
+func capsuleVsBoxDistance(c *capsule, other *box) float64 {
+	// Capsules are not polyhedrons, so separating axis test is not ideal here.
+	meshDist := capsuleVsMeshDistance(c, other.toMesh())
```

Comment:
I should probably swap this over to the Separating Axis Theorum as well.

There were *no* examples I could find of using the SAT for capsule vs box, and implementations I did find tended to convert capsules into triangles and do a mesh collision, hence why meshes are partially implemented in this PR. Since that's how it's commonly done, I began with the assumption it would be faster than the SAT.

Turns out it's slower than the SAT by quite a bit, so once I could compare I swapped back over. I'll swap this as well.